File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/c/p4c0j0n3s.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

int MIN_LENGTH = 8;
int MAX_LENGTH = 16;

int main(int argc, char *argv[]){

    static const char lower_case_letters[] = "abcdefghijklmnopqrstuvwxyz";
    static const char upper_case_letters[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    static const char *numbers = "1234567890";
    static const char *symbols = "!$&()=~,.-;:_";

    char characters[80] = {};

    int length_characters = 0;

    int length_password = 0;

    int i, len;
    time_t t;

    if (argc == 1) {
        printf("\nUsage: %s num [OPTIONS] \n\n\
Get a multicharacter password (Default: lower case letters) \n\n\
- num: number of characters of the password. \n\n\
- OPTIONS: \n\
    U: Upper case letters \n\
    n: Numbers \n\
    s: Symbol s \n\
    ", argv[0]);
        exit(0);
    }

    len = atoi(argv[1]);

    length_password = (len >= MIN_LENGTH && len <= MAX_LENGTH) ? len : 0;

    if (length_password == 0) {
        printf("The password length must be %d and %d", MIN_LENGTH, MAX_LENGTH);
        exit(0);
    }

    strcat(characters, lower_case_letters);
    length_characters += strlen(lower_case_letters);;
    
    if (argc == 3) {

        if ( strchr(argv[2], 'U') != NULL ) {
            length_characters += strlen(upper_case_letters);;
            strcat(characters, upper_case_letters); 
        }
         
        if (strchr(argv[2], 'n') != NULL ) {
            length_characters += strlen(numbers);
            strcat(characters, numbers);
        }
        
        if (strchr(argv[2], 's') != NULL ) {
            length_characters += strlen(symbols);
            strcat(characters, symbols);        }
    }

    srand((unsigned) time(&t));

    printf("Password --> ");
    for (i=0; i < length_password; i++){
        printf("%c", characters[rand() % length_characters]);
    }   
}

File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/rust/p4c0j0n3s.rs
use std::env;
use std::process::exit;
use rand::Rng;

fn main(){

    const MIN_LENGTH: usize = 8;
    const MAX_LENGTH: usize = 16;

    let lower_case_letters: &'static str = "abcdefghijklmnopqrstuvwxyz";
    let upper_case_letters: &'static str = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    let numbers: &'static str = "1234567890";
    let symbols: &'static str = "!$&()=~,.-;:_";

    let mut characters = String::with_capacity(80);

    let args: Vec<String> = env::args().collect();

    if args.len() == 1 {
            println!("\nUsage: {} num [OPTIONS] \n\n\
            Get a multicharacter password (Default: lower case letters) \n\n\
            - num: number of characters of the password. \n\n\
            - OPTIONS: \n    U: Upper case letters \n    n: Numbers \n    s: Symbol
            ", args[0]);
            exit(0);    
    } 

    let len = match &args[1].parse::<usize>() {
        Ok(n) => {
            (*n).try_into().unwrap()
        },
        Err(_) => {
            eprintln!("The first argument has to be a number between {} and {}", MIN_LENGTH, MAX_LENGTH);
            return;
        },
    };

    let length_password = if len >= MIN_LENGTH && len <= MAX_LENGTH { len } else { 0 };

    if length_password == 0 {
        println!("The first argument has to be a number between {} and {}", MIN_LENGTH, MAX_LENGTH);
    } else {

        characters.push_str(lower_case_letters);
    
        if args.len() == 3 {
            if args[2].contains("U"){ characters.push_str(upper_case_letters); }
            if args[2].contains("n"){ characters.push_str(numbers); }
            if args[2].contains("s"){ characters.push_str(symbols); }
        }

        let mut rng = rand::thread_rng();
        let letters: Vec<char> = characters.chars().collect();

        print!("Password --> ");

        for _i in 0..length_password {
            print!("{}", letters[ rng.gen::<usize>() % characters.len() ]);
        }
    }
}